### PR TITLE
ARROW-10679: [Rust] [DataFusion] Implement CASE WHEN physical expression

### PR DESCRIPTION
**Pull request by andygrove**
This PR implements the physical expression for the SQL CASE expression.

The CASE expression is similar to a series of nested if/else and there are two forms that can be used. 

The first form consists of a series of boolean "when" expressions with corresponding "then" expressions and an optional "else" expression.

```sql
CASE WHEN condition THEN result
     [WHEN ...]
     [ELSE result]
END
```

The second form uses a base expression and then a series of "when" clauses that match on a literal value.

```sql
CASE expression
    WHEN value THEN result
    [WHEN ...]
    [ELSE result]
END
```

The implementation is quite simple/naive and I have not attempted to optimize this yet. I would prefer to get this plumbed through and tested with TPC-H before optimizing.

### CODE REVIEW

**Comment by andygrove** [comment]
@jorgecarleitao @alamb I could use some advice on the best way to make this new code support all array types. In the past, I have gone with macros because I know how to do that, but perhaps I should be trying to use generics here?

There may also be some patterns in here that we want to extract and implement as kernels in the arrow crate. In particular, I would like an `if` primitive so that I can do something like:

```rust
let new_array = boolean_array.ifThenElse(array1, array2);
```

If I had this it would simply this code quite a lot.

**Comment by jorgecarleitao** [comment]
Cool!

IMO we want to "take" from different arrays depending on a set of boolean conditions. My recommendation is to use  `MutableDataArray` API that we also use in the Join and that was though through exactly for this use-case.

Something like 

```rust
let boolean_conditions: Vec<BooleanArray> = <args>
let first_non_null: Vec<usize> = first_true(boolean_conditions)  // index from 0 to boolean_conditions.len() (inclusive, last represents the "else")

let capacity = boolean_conditions[0].len();
let mutable = MutableDataArray::new(arrays.iter().map(|array| array.data_ref()).collect::<Vec<_>>(), capacity);

for row in boolean_conditions[0].len() {
    for index in first_non_null {
        if index == boolean_conditions.len() {
            // the else case
            mutable.extend_null(index); // this function is still not available in master, but I have a branch with it
        } else {
            // take the slice `[row,row+1[` from array "index".
            mutable.extend_from_slice(index, row, row + 1);
        }
    }
}

return make_array(Arc::new(mutable.freeze()));
```

the advantage of this is that it works for all types except struct.

**Comment by andygrove** [comment]
@jorgecarleitao I went ahead and fully implemented this the inefficient way for now. It will take me some time to get familiar with the MutableArrayData approach so I would like to follow up with a separate PR to optimize this once it is plumbed through and fully tested.

File: rust/datafusion/src/physical_plan/mod.rs [hunk 1]
```diff
@@ -126,10 +126,10 @@ impl ColumnarValue {
         }
     }
 
-    fn into_array(self, batch: &RecordBatch) -> ArrayRef {
+    fn into_array(self, num_rows: usize) -> ArrayRef {
```
Comment (jorgecarleitao) [inline rust/datafusion/src/physical_plan/mod.rs:129]: good. I had the same concern when reviewing this the other day.

File: rust/datafusion/src/physical_plan/expressions.rs [hunk 2]
```diff
@@ -1697,6 +1700,409 @@ pub fn is_not_null(arg: Arc<dyn PhysicalExpr>) -> Result<Arc<dyn PhysicalExpr>>
...
+        for row in 0..when_value.len() {
+            if when_value.is_valid(row) && base_value.is_valid(row) {
+                builder.append_value(when_value.value(row) == base_value.value(row))?;
+            } else {
+                builder.append_null()?;
+            }
+        }
+        Ok(builder.finish())
+    }};
+}
```
Comment (jorgecarleitao) [inline rust/datafusion/src/physical_plan/expressions.rs:1956]: Doesn't this exist as `arrow::compute::kernels::comparison::eq`?

**Review by jorgecarleitao — APPROVED** [review]
Went through this. Looks good to me. 

I left a comment in one of the functions.

A potential concern is that there is some code around that IMO can be heavily simplified and optimized, but IMO having the `case` is more important than those details, so, :shipit:

File: rust/datafusion/src/physical_plan/expressions.rs (hunk 2 above)
Comment (andygrove) [reply rust/datafusion/src/physical_plan/expressions.rs:1956]: I did try and use that and ran into an issue. I will merge this and then create a follow-up PR so we can discuss that. I may need some help. I am on vacation this week by the way, hence I have some time to focus on this. My goal is to have the complete TPC-H suite working from SQL this week without any shortcuts.

File: rust/datafusion/src/physical_plan/mod.rs (hunk 1 above)
Comment (alamb) [reply rust/datafusion/src/physical_plan/mod.rs:129]: yeah this is a nice change

**Review by alamb — COMMENTED** [review]
As @andygrove  says, this code has many further optimization opportunities, but as an initial implementation to get things working it looks good. Thank you @andygrove  and @jorgecarleitao

**Pull request closed without merging**